### DISCUSSION TITLE
ImportError: cannot import name 'rcParams' from 'matplotlib.axes' error with PyInstaller v6.0.0 onwards

### DISCUSSION THREAD

**Discussion by olumide-x** [pyinstaller/pyinstaller#8266] [Help]
When I run a binary packaged with PyInstaller v6.0.0 onwards I get this error:

```
Traceback (most recent call last):
  File "src/Application.py", line 10, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 391, in exec_module
  ...
  (omitted own modules)
  ...
  File "matplotlib/figure.py", line 43, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 391, in exec_module
  File "matplotlib/projections/__init__.py", line 58, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 391, in exec_module
  File "mpl_toolkits/mplot3d/__init__.py", line 1, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 391, in exec_module
  File "mpl_toolkits/mplot3d/axes3d.py", line 34, in <module>
ImportError: cannot import name 'rcParams' from 'matplotlib.axes' (/path/to/Application/_internal/matplotlib/axes/__init__.pyc)
[58415] Failed to execute script 'Application' due to unhandled exception!
```
Is this a known issue?

**Comment by rokm** [comment]
Hmm, no. Can you provide a minimal example that triggers this (i.e., what projection are you using with the figure)?

**Reply by olumide-x** [reply]
The problem happens on this example:
https://matplotlib.org/3.4.3/gallery/user_interfaces/pylab_with_gtk_sgskip.html

```
pyi-makespec --noconsole matplotlib_example.py
pyinstaller --clean --noconfirm matplotlib_example.spec
```

**Reply by rokm** [reply]
Thanks, I'll take a look at it.

**Reply by rokm** [reply]
Hmmm, I cannot seem to reproduce this. What version of `matplotlib` are you using?

**Reply by olumide-x** [reply]
```
python3 -c "import matplotlib; print('matplotlib: {}'.format(matplotlib.__version__))"
matplotlib: 3.7.2
```
OS: Ubuntu 22.04 LTS

**Reply by rokm** [reply]
In that case, can you open `build/<progamname>/Analysis-00.toc` and check where `matplotlib` files were collected from? For example, `matplotlib/__init__.py` and `mpl_toolkits/mplot3d/axes3d.py`.

In 3.7.2, the `mpl_toolkits/mplot3d/axes3d.py` does not attempt to import `rcParams` from `matplotlib`: https://github.com/matplotlib/matplotlib/blob/v3.7.2/lib/mpl_toolkits/mplot3d/axes3d.py#L13-L39

But it did in 3.5.x series: https://github.com/matplotlib/matplotlib/blob/v3.5.3/lib/mpl_toolkits/mplot3d/axes3d.py#L34
The error line matches the one from your error traceback.

Plus, Ubuntu 22.04 LTS seems to have matplotlib 3.5.x packaged in its repositories. So my guess would be that somehow two versions of matplotlib are getting mixed up during the build.

**Reply by olumide-x** [reply]
```
grep -n matplotlib Analysis-00.toc
580:  ('matplotlib.pyplot',
581:   '/home/olumide/.local/lib/python3.10/site-packages/matplotlib/pyplot.py',
583:  ('matplotlib.patheffects',
584:   '/home/olumide/.local/lib/python3.10/site-packages/matplotlib/patheffects.py',
586:  ('matplotlib.path',
587:   '/home/olumide/.local/lib/python3.10/site-packages/matplotlib/path.py',
589:  ('matplotlib.hatch',
590:   '/home/olumide/.local/lib/python3.10/site-packages/matplotlib/hatch.py',
592:  ('matplotlib.bezier',
593:   '/home/olumide/.local/lib/python3.10/site-packages/matplotlib/bezier.py',
595:  ('matplotlib.transforms',
596:   '/home/olumide/.local/lib/python3.10/site-packages/matplotlib/transforms.py',
997:  ('matplotlib.backends',
998:   '/home/olumide/.local/lib/python3.10/site-packages/matplotlib/backends/__init__.py',
1000:  ('matplotlib.backends.backend_gtk3agg',
1001:   '/home/olumide/.local/lib/python3.10/site-packages/matplotlib/backends/backend_gtk3agg.py',
1006:  ('matplotlib.backends.backend_gtk3',
1007:   '/home/olumide/.local/lib/python3.10/site-packages/matplotlib/backends/backend_gtk3.py',
1009:  ('matplotlib.backend_tools',
1010:   '/home/olumide/.local/lib/python3.10/site-packages/matplotlib/backend_tools.py',
1013:  ('matplotlib.backends._backend_gtk',
1014:   '/home/olumide/.local/lib/python3.10/site-packages/matplotlib/backends/_backend_gtk.py',
1016:  ('matplotlib.backends.backend_webagg',
1017:   '/home/olumide/.local/lib/python3.10/site-packages/matplotlib/backends/backend_webagg.py',
1019:  ('matplotlib.backends.backend_webagg_core',
1020:   '/home/olumide/.local/lib/python3.10/site-packages/matplotlib/backends/backend_webagg_core.py',
1234:  ('matplotlib.backends.backend_agg',
1235:   '/home/olumide/.local/lib/python3.10/site-packages/matplotlib/backends/backend_agg.py',
1237:  ('matplotlib.mathtext',
1238:   '/home/olumide/.local/lib/python3.10/site-packages/matplotlib/mathtext.py',
...
```

Also

```
apt list --installed matplotlib
Listing... Done
```

**Reply by bwoodsend** [reply]
`apt list --installed python3-matplotlib` would be the proper test but it looks like your matplotlib is installed into the system environment using pip but with the `--user` flag?

**Reply by rokm** [reply]
What about `mpl_toolkits/mplot3d/axes3d.py`? I.e., `grep -n mpl_toolkits Analysis-00.toc`?

**Reply by olumide-x** [reply]
```
grep -n mpl_toolkits Analysis-00.toc
4229:  ('mpl_toolkits.mplot3d',
4230:   '/usr/lib/python3/dist-packages/mpl_toolkits/mplot3d/__init__.py',
4232:  ('mpl_toolkits',
4233:   '/usr/lib/python3/dist-packages/mpl_toolkits/__init__.py',
4235:  ('mpl_toolkits.mplot3d.axes3d',
4236:   '/usr/lib/python3/dist-packages/mpl_toolkits/mplot3d/axes3d.py',
4349:  ('mpl_toolkits.mplot3d.axis3d',
4350:   '/usr/lib/python3/dist-packages/mpl_toolkits/mplot3d/axis3d.py',
4352:  ('mpl_toolkits.mplot3d.art3d',
4353:   '/usr/lib/python3/dist-packages/mpl_toolkits/mplot3d/art3d.py',
4355:  ('mpl_toolkits.mplot3d.proj3d',
4356:   '/usr/lib/python3/dist-packages/mpl_toolkits/mplot3d/proj3d.py',
```

**Reply by rokm** [reply]
Yeah, so pip-installed `matplitlib` from your user site and `mpl_toolkits` from the deb package are being mixed up, and they are incompatible. The mix-up could be caused by the way PyInstaller's modulegraph searches for packages/modules, and you would avoid it by using either virtual environment (with pip-installed components) or only your system-provided environment (i.e., just deb-installed python packages, without any pip-installed components in user site).

**Reply by olumide-x** [reply]
Thanks for the workaround. Will future versions of PyInstaller address this issue?

**Reply by bwoodsend** [reply]
I don't think there's much we can do about it. There's a reason why pip tells you not to use it in a system environment.

**Reply by rokm** [reply]
> Thanks for the workaround. Will future versions of PyInstaller address this issue?

Unlikely, because as far as I'm concerned, user site should be disabled on linux. Perhaps once we rewrite/replace the modulegraph, this will sort itself out.

**Reply by olumide-x** [reply]
Thanks. Can I just mention that your support is stellar by any standard.

**Comment by olumide-x** [comment]
> `apt list --installed python3-matplotlib` would be the proper test but it looks like your matplotlib is installed into the system environment using pip but with the `--user` flag?

```
apt list --installed python3-matplotlib
Listing... Done
python3-matplotlib/jammy,now 3.5.1-2build1 amd64 [installed,automatic]
```
